### ISSUE TITLE
Unable to resolve Route 53 Alias Records

### ISSUE THREAD

**Issue by wazoo** [coredns/coredns#3012]
I am using the following: CoreDNS-1.5.0 - linux/amd64, go1.12.2, e3f9a80

I am using CoreDNS on a regular linux machine in a docker container to provide some DNS routing and split horizon for a hybrid AWS and on-prem infrastructure.  Normal DNS resolution works great however when I try to resolve an [AWS Alias Record](https://aws.amazon.com/premiumsupport/knowledge-center/route-53-create-alias-records/) I get NXDOMAIN.

I have tried adding a forward and fallthrough to the zone config (shown below) and still get NXDOMAIN but querying the VPC resolver for the same record works fine.

I have all of the zones configured as follows:

```
(base) {
  log
  errors
}
(cache) {
  cache 30 {
    success 500 90
    denial 500
  }
}
sub.domain.com.:53 {
  import base
  import cache
  route53 sub.domain.com.:<private zone ID> sub.domain.com.io.:<public zone ID>  {
      import account-credential-file
  }
}
```
If I have a record (as an example) `record.sub.domain.com` that is an A record or a CNAME resolution works as expected and private records override public records.  

If I have a record (as an example) `alias.sub.domain.com` I receive the following:

```
2019-07-16T20:52:55.952Z [INFO] 172.17.0.1:35996 - 9166 "A IN alias.sub.domain.com. udp 48 false 4096" NXDOMAIN qr,aa,rd 134 0.0002356s
```

**Comment by wazoo** [comment]
I did some additional testing and I was able to get it to work with the following config:

```
(base) {
  log
  errors
}
(cache) {
  cache 30 {
    success 500 90
    denial 500
  }
}
sub.domain.com.:53 {
  import base
  import cache
  route53 sub.domain.com.:<private zone ID> sub.domain.com.io.:<public zone ID>  {
      import account-credential-file
      fallthrough
  }
  forward <vpc-resolver>:53
}
```

I am not sure if this is just missing from the documentation or I missread it but I am not sure why I need to specifically forward every zone.  I don't have a root forwarder configured on these containers, if I did that would this have the same effect?

**Comment by sharmaansh21** [comment]
@wazoo I will debug into it, is it possible for you to provide more details. or maybe you can send it to me sharmaanshul2102@gmail.com if you don't want to share here. 

Exact config will be helpful for debugging.

**Comment by ykhr53** [comment]
Similar behavior is happened on my side too.

- CoreDNS-1.6.5 (also 1.6.4)
- go version go1.12.5 darwin/amd64 (also go1.12.5 linux/amd64)

I can reproduce with a simple Corefile like below.  

```corefile
private.org. {
    route53 private.org.:Z123EXAMPLEAA
    log
}
```

```txt
[INFO] 127.0.0.1:63184 - 13717 "A IN alias.private.org. udp 46 false 4096" NXDOMAIN qr,aa,rd 133 0.00015149s
[INFO] 127.0.0.1:65272 - 14972 "A IN www.private.org. udp 44 false 4096" NOERROR qr,aa,rd 246 0.00011294s
[INFO] 127.0.0.1:51783 - 15085 "SOA IN private.org. udp 40 false 4096" NOERROR qr,aa,rd 309 0.000080274s
```

When NXDOMAIN is returned, I get an answer with AA bit and private hosted zone's SOA in Auth section.
So it seems the answer is returned from CoreDNS correctly.
Also I can receive properly other than alias records.
It will happen both Private and Public Hosted Zone.

No one doesn't mentioned about it, so I'm a little confused.
If I overlook or misconfigure, please point it.
For now I am not using an alias record, I just noticed and reported it 😊
It's just like a FYI, I'm looking into the code anyway.

Thanks,

**Comment by miekg** [comment]
Is it possible to publicly test this behaviour?

What should be returned when things works? An A or AAAA record instead of NXDOMAIN?

**Comment by ykhr53** [comment]
Maybe I can share a IP address for a short time for testing.
And it's correct, I'm expecting to get A or AAAA record. (If this plugin supports Alias record.)

I looked into these return values for the query which has an Alias record.
https://github.com/coredns/coredns/blob/e23a34abb439c8169f90a2fc431e0ba0019d6a50/plugin/route53/route53.go#L122

I got no answer and `result` as `1` (= NXDOMAIN).
So it seems an Alias record can't be resolved by `hostedZone.z.Lookup(ctx, state, qname)` (line 122).
I am not sure how to handle and store Alias records in this plugin, but I think we can't get an Alias Target's actual IP address by API.
I think we need to use `fallthrough` and `forward` to VPC resolver.

I just want to clarify whether this plugin supports [Alias records](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-choosing-alias-non-alias.html) or not.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Unable to res..." ]
>Maybe I can share a IP address for a short time for testing.
>And it's correct, I'm expecting to get A or AAAA record. (If this plugin supports Alias record.)
>
>I looked into this return values for the query which has an Alias record.
>https://github.com/coredns/coredns/blob/e23a34abb439c8169f90a2fc431e0ba0019d6a50/plugin/route53/route53.go#L122
>
>I got no answer and `result` as `1` (= NXDOMAIN).
>So it seems an Alias record can't be resolved by `hostedZone.z.Lookup(ctx, state, qname)` (line 122).
>I am not sure how to handle and store Alias records in this plugin, but I think we can't get an Alias Target's actual IP address by API.
>I think we need to use `fallthrough` and `forward` to VPC resolver.
>
>I just want to clarify whether this plugin supports [Alias records](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-choosing-alias-non-alias.html) or not.

That I don't know. Maybe @dilyevsky knows?

**Comment by ilijamt** [comment]
Yeah I'm also interested in this, the problem is that the current implementation of route53 doesn't support aliases, as it's not part of the DNS standard. 

There is a PR that tried to implement it but was closed, the PR is https://github.com/coredns/coredns/pull/2193

This needs to be evaluated and seen if it can be implemented in the current route53 plugin.

I started working on it a while ago, but kinda got sidetracked and never did finish it in https://github.com/ilijamt/coredns/tree/route53-alias

I got to a point where it fetches the data from the zone file, now I needed to make sure that we replace the old records with the new ones and thats where I stopped, as I need to do actual DNS queries to get the A/AAAA records of Amazon DNS server, and in the case where we are hosting the DNS from our zones.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Unable to res..." ]
>Yeah I'm also interested in this, the problem is that the current implementation of route53 doesn't support aliases, as it's not part of the standard.

What 'standard' you mean? The DNS one? If so, we have to be careful here with 
what we allow. Note the dnsop within the IETF is working on a generic solution 
here. HTTPSVC record I believe it's called atm.

**Comment by yongtang** [comment]
Route53's alias target is not a DNS standard.

In a more or less similar way, there was a ALIAS concept for RFC at one point: https://tools.ietf.org/html/rfc2219. Though I believe RFC2219 is not in standard?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Unable to res..." ]
>Route53's alias target is not a DNS standard.
>
>In a more or less similar way, there was a ALIAS concept for RFC at one point: https://tools.ietf.org/html/rfc2219. Though I believe RFC2219 is not in standard?

and older draft is here: https://tools.ietf.org/id/draft-nygren-httpbis-httpssvc-00.html

**Comment by ykhr53** [comment]
Alias record concept itself is maybe not standard (tbh I'm not sure), but as the name resolution perspective, route 53 authoritative name server just returns A/AAAA record.
In the PR #3442, the plugin queries to route 53 name servers to get the A/AAAA record.
It is just my opinion though, I think it is acceptable.

**Comment by ilijamt** [comment]
Should be acceptable it also help us maintain health checks if any by Route 53.
Still waiting on comments on the PR hopefully we will get some more info soon so we can proceed.

**Comment by ilijamt** [comment]
@miekg even the API will not tell you how to resolve it it will just tell you that it's an alias and pointing to a different fqdn. And based on the configuration Route 53 can add or remove IP's from the dns entry. I see it happen quite often especially with Application Load Balancers, and/or configured healthchecks in AWS. So the only way to know for sure is to let the AWS authoritative DNS resolve it for us, as we don't have access to the zone that hosts the aliases.
